plugin/pkg/scheduler/core/scheduling_queue.go

```diff
@@ -15,47 +15,564 @@ limitations under the License.
+	nominatedNodeName, ok := pod.Annotations[NominatedNodeAnnotationKey]
+	if !ok {
+		return ""
+	}
+	return nominatedNodeName
+}
+
+// Add adds a pod to the unschedulable pods.
+func (u *UnschedulablePodsMap) Add(pod *v1.Pod) {
+	nominatedNodeName := NominatedNodeName(pod)
+	if _, exists := u.pods[nominatedNodeName]; !exists {
+		u.pods[nominatedNodeName] = make(map[string]*v1.Pod)
+	}
+	u.pods[nominatedNodeName][util.GetPodFullName(pod)] = pod
+}
+
+// Delete deletes a pod from the unschedulable pods.
+func (u *UnschedulablePodsMap) Delete(pod *v1.Pod) {
+	nominatedNodeName := NominatedNodeName(pod)
+	if pods, ok := u.pods[nominatedNodeName]; ok {
+		delete(pods, util.GetPodFullName(pod))
+		if len(u.pods[nominatedNodeName]) == 0 {
+			delete(u.pods, nominatedNodeName)
+		}
+	}
+}
+
+// Exists returns true if the given pod is found in the map.
+func (u *UnschedulablePodsMap) Exists(pod *v1.Pod) bool {
+	nominatedNodeName := NominatedNodeName(pod)
+	if pods, ok := u.pods[nominatedNodeName]; ok {
+		_, exists := pods[util.GetPodFullName(pod)]
+		return exists
+	}
+	return false
+}
+
+// GetPodsForNode returns a list of unschedulable pods whose NominatedNodeNames
+// are equal to the given nodeName.
+func (u *UnschedulablePodsMap) GetPodsForNode(nodeName string) map[string]*v1.Pod {
+	if pods, ok := u.pods[nodeName]; ok {
+		return pods
+	}
+	return nil
+}
+
+// Clear removes all the entries from the unschedulable map.
+func (u *UnschedulablePodsMap) Clear() {
+	u.pods = make(map[string]map[string]*v1.Pod)
+}
+
+// newUnschedulablePodsMap initializes a new object of UnschedulablePodsMap.
+func newUnschedulablePodsMap() *UnschedulablePodsMap {
+	return &UnschedulablePodsMap{
+		pods: make(map[string]map[string]*v1.Pod),
+	}
+}
+
+// Below is the implementation of the a heap. The logic is pretty much the same
+// as cache.heap, however, this heap does not perform synchronization. It leaves
+// synchronization to the SchedulingQueue.
+
+type LessFunc func(interface{}, interface{}) bool
+type KeyFunc func(obj interface{}) (string, error)
+
+type heapItem struct {
+	obj   interface{} // The object which is stored in the heap.
+	index int         // The index of the object's key in the Heap.queue.
+}
+
+type itemKeyValue struct {
+	key string
+	obj interface{}
+}
+
+// heapData is an internal struct that implements the standard heap interface
+// and keeps the data stored in the heap.
+type heapData struct {
+	// items is a map from key of the objects to the objects and their index.
+	// We depend on the property that items in the map are in the queue and vice versa.
+	items map[string]*heapItem
+	// queue implements a heap data structure and keeps the order of elements
+	// according to the heap invariant. The queue keeps the keys of objects stored
+	// in "items".
+	queue []string
+
+	// keyFunc is used to make the key used for queued item insertion and retrieval, and
+	// should be deterministic.
+	keyFunc KeyFunc
```

Comment:
looks like this is owned by `Heap`. `heapData` isn't using it.

Reply:
That's a valid point; `headData` does not use `keyFunc`. This is in `heapData` because `heapData` holds all the data that `Heap` needs.